File: route.cpp
```cpp
// Finds the shortest path between two nodes in a graph using Dijkstra's algorithm.
// By: Jonathan Mitelman
#include <iostream>
#include <vector>
#include <queue>
#include <map>
#include <limits>

using namespace std;

typedef pair<int, string> Pair;

pair<vector<string>, int> shortest_path_dijkstra(map<string, map<string, int>>& graph, string start, string end) {
    map<string, int> distances;
    map<string, string> predecessors;

    for (auto const& [node, neighbors] : graph) {
        distances[node] = numeric_limits<int>::max();
        predecessors[node] = "";
    }

    distances[start] = 0;

    priority_queue<Pair, vector<Pair>, greater<Pair>> pq;
    pq.push({0, start});

    while (!pq.empty()) {
        int dist = pq.top().first;
        string current_node = pq.top().second;
        pq.pop();

        if (dist > distances[current_node]) {
            continue;
        }

        if (current_node == end) {
            break;
        }

        for (auto const& [neighbor, weight] : graph[current_node]) {
            int new_dist = dist + weight;

            if (new_dist < distances[neighbor]) {
                distances[neighbor] = new_dist;
                predecessors[neighbor] = current_node;
                pq.push({new_dist, neighbor});
            }
        }
    }

    vector<string> path;
    string current = end;
    while (current != "") {
        path.insert(path.begin(), current);
        if(predecessors[current] == "" && current != start){
            return {vector<string>(), numeric_limits<int>::max()};
        }
        current = predecessors[current];
    }

    if (distances[end] == numeric_limits<int>::max()) {
        return {vector<string>(), numeric_limits<int>::max()};
    } else {
        return {path, distances[end]};
    }
}

int main() {
    map<string, map<string, int>> graph = {
        {"A", {{"B", 5}, {"C", 2}}},
        {"B", {{"A", 5}, {"D", 1}}},
        {"C", {{"A", 2}, {"D", 8}}},
        {"D", {{"B", 1}, {"C", 8}, {"E", 7}}},
        {"E", {{"D", 7}}}
    };

    string start_node = "A";
    string end_node = "E";

    pair<vector<string>, int> result = shortest_path_dijkstra(graph, start_node, end_node);
    vector<string> path = result.first;
    int cost = result.second;

    if (!path.empty()) {
        cout << "Shortest path from " << start_node << " to " << end_node << ": ";
        for (const string& node : path) {
            cout << node << " ";
        }
        cout << endl;
        cout << "Total cost: " << cost << endl;
    } else {
        cout << "No path found from " << start_node << " to " << end_node << endl;
    }

    return 0;
}
```

File: route.js
```javascript
/** Finds the shortest path from start to end in a graph, visiting all nodes in must_visit, with a maximum allowed total distance.
By: Jonathan Mitelman */
function shortestPathWithConstraints(graph, start, end, mustVisit, maxDistance) {
    function getNeighbors(node) {
        return graph[node] || {};
    }

    let shortestPath = null;
    let shortestDistance = Infinity;

    function backtrack(currentNode, visited, currentPath, currentDistance) {
        if (
            Array.from(mustVisit).every(node => visited.has(node)) &&
            currentNode === end
        ) {
            if (currentDistance < shortestDistance) {
                shortestDistance = currentDistance;
                shortestPath = [...currentPath];
            }
            return;
        }

        if (currentDistance >= shortestDistance || currentDistance > maxDistance) {
            return;
        }

        const neighbors = getNeighbors(currentNode);
        for (const neighbor in neighbors) {
            const distance = neighbors[neighbor];
            if (!visited.has(neighbor)) {
                visited.add(neighbor);
                currentPath.push(neighbor);
                backtrack(neighbor, visited, currentPath, currentDistance + distance);
                currentPath.pop();
                visited.delete(neighbor);
            }
        }
    }

    const visited = new Set([start]);
    backtrack(start, visited, [start], 0);

    return [shortestPath, shortestDistance];
}

if (typeof require !== 'undefined' && require.main === module) {
    const graph = {
        'A': { 'B': 5, 'C': 2 },
        'B': { 'A': 5, 'D': 1 },
        'C': { 'A': 2, 'D': 8 },
        'D': { 'B': 1, 'C': 8, 'E': 7 },
        'E': { 'D': 7 }
    };

    let path, distance;

    path, distance = shortestPathWithConstraints(graph, 'A', 'E', new Set(), 20);
    console.log(`Path: ${path}, Distance: ${distance}`);

    path, distance = shortestPathWithConstraints(graph, 'A', 'E', new Set(['B']), 30);
    console.log(`Path: ${path}, Distance: ${distance}`);

    path, distance = shortestPathWithConstraints(graph, 'A', 'E', new Set(['B']), 10);
    console.log(`Path: ${path}, Distance: ${distance}`);

    path, distance = shortestPathWithConstraints(graph, 'A', 'E', new Set(['B', 'C']), 30);
    console.log(`Path: ${path}, Distance: ${distance}`);

    path, distance = shortestPathWithConstraints(graph, 'A', 'F', new Set(['B']), 30);
    console.log(`Path: ${path}, Distance: ${distance}`);

    path, distance = shortestPathWithConstraints(graph, 'A', 'E', new Set(), 15);
    console.log(`Path: ${path}, Distance: ${distance}`);

    const graph2 = {
        'A': { 'B': 2, 'C': 5 },
        'B': { 'A': 2, 'D': 3, 'E': 7 },
        'C': { 'A': 5, 'F': 8 },
        'D': { 'B': 3, 'G': 4 },
        'E': { 'B': 7, 'H': 6 },
        'F': { 'C': 8, 'I': 9 },
        'G': { 'D': 4, 'J': 5 },
        'H': { 'E': 6, 'K': 10 },
        'I': { 'F': 9, 'L': 11 },
        'J': { 'G': 5 },
        'K': { 'H': 10 },
        'L': { 'I': 11 }
    };

    path, distance = shortestPathWithConstraints(graph2, 'A', 'J', new Set(['B', 'D']), 20);
    console.log(`Path: ${path}, Distance: ${distance}`);
}
```